File: 2023/kotlin/src/Day25.kt
typealias Component = String
typealias Subgroup = MutableSet<Component>
typealias Subgroups = MutableSet<Subgroup>

fun Subgroups.find(c: Component) = first { c in it }
fun Subgroups.merge(a: Subgroup, b: Subgroup) {
    if (a === b) return;
    a.addAll(b)
    removeIf { it === b }
}

// https://en.wikipedia.org/wiki/Karger%27s_algorithm
private fun part1(input: String): Int {
    val lines = input.lines().map {
        it.split(':', ' ').filterNot(String::isBlank)
    }

    val components = lines.flatten().toSet()
    val connections = buildSet {
        lines.forEach { line ->
            addAll(line.drop(1).map { line.first() to it })
        }
    }

    while (true) {
        val subgroups = components.map { mutableSetOf(it) }.toMutableSet()
        val randomConnection = connections.shuffled().iterator()

        while (subgroups.size > 2) {
            val (component1, component2) = randomConnection.next()

            val subgroup1 = subgroups.find(component1)
            val subgroup2 = subgroups.find(component2)

            subgroups.merge(subgroup1, subgroup2)
        }

        val wires = connections.count { (a, b) -> subgroups.find(a) != subgroups.find(b) }
        if (wires > 3) continue

        return subgroups.map { it.size }.reduce(Int::times)
    }
}

fun main() {
    val input = readInput("Day25")
    part1(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day25.swift
typealias Component = String
typealias Subgroup = Set<Component>
typealias Subgroups = Set<Subgroup>

extension Subgroups {
  func find(component: Component) -> Subgroup {
    return first(where: { $0.contains(component) })!
  }

  mutating func merge(_ a: Subgroup, _ b: Subgroup) {
    if a == b { return }
    remove(a)
    remove(b)
    insert(a.union(b))
  }
}

protocol PairConvertible {
  associatedtype A
  associatedtype B
  init(_ pair: (A, B))
}

struct PairStruct<A, B>: Hashable, PairConvertible where A: Hashable, B: Hashable {
  let a: A
  let b: B

  init(_ pair: (A, B)) {
    self.a = pair.0
    self.b = pair.1
  }
}

struct Day25: AdventDay {
  var input: String

  // https://en.wikipedia.org/wiki/Karger%27s_algorithm
  func part1() -> Any {
    let lines = input.lines().map {
      $0.split(whereSeparator: { $0 == " " || $0 == ":" })
        .map { "\($0)".trimmingCharacters(in: .whitespaces) }
        .filter { !$0.isEmpty }
    }

    let components = Set(lines.joined())
    let connections = Set(
      lines.flatMap { line in
        line.dropFirst(1).map { PairStruct((line[0], "\($0)")) }
      })

    while true {
      var subgroups = Set(components.map { Set([$0]) })
      var randomConnection = connections.shuffled().makeIterator()

      while subgroups.count > 2 {
        let components = randomConnection.next()!

        let subgroup1 = subgroups.find(component: components.a)
        let subgroup2 = subgroups.find(component: components.b)

        subgroups.merge(subgroup1, subgroup2)
      }

      let wires =
        connections
        .filter { components in
          subgroups.find(component: components.a) != subgroups.find(component: components.b)
        }
        .count
      if wires > 3 { continue }

      return subgroups.map { $0.count }.reduce(1, *)
    }
  }
}
